lib/ruby-mpd/plugins/command_list.rb

```diff
@@ -0,0 +1,177 @@
+    # * +:playlists+ - an array of Playlist instances from the results
+    #
+    # Note that each supported command has no return value inside the block.
+    # Instead, the block itself returns the array of results.
+    #
+    # @param [Symbol] response_type the type of responses desired.
+    # @return [nil] default behavior.
+    # @return [Array] if +response_type+ is +:values+, +:hashes+, +:songs+, or +:playlists+.
+    # @return [Hash] if +response_type+ is +:hash+.
+    #
+    # @example Simple batched control commands
+    #   @mpd.command_list do
+    #     stop
+    #     shuffle
+    #     save "shuffled"
+    #   end
+    #
+    # @example Adding songs to the queue, ignoring the response
+    #   @mpd.command_list do
+    #     my_songs.each do |song|
+    #       add(song)
+    #     end
+    #   end
+    #
+    # @example Adding songs to the queue and getting the song ids
+    #   ids = @mpd.command_list(:values){ my_songs.each{ |song| addid(song) } }
+    #   #=> [799,800,801,802,803]
+    #
+    #   ids = @mpd.command_list(:hashes){ my_songs.each{ |song| addid(song) } }
+    #   #=> [ {:id=>809}, {:id=>810}, {:id=>811}, {:id=>812}, {:id=>813} ]
+    #
+    #   ids = @mpd.command_list(:hash){   my_songs.each{ |song| addid(song) } }
+    #   #=> { :id=>[804,805,806,807,808] }
+    #
+    # @example Finding songs matching various genres
+    #   songs = @mpd.command_list(:songs) do
+    #     where genre:'Alternative Rock'
+    #     where genre:'Alt. Rock'
+    #     where genre:'alt-rock'
+    #   end
+    #
+    # @see CommandList::Commands CommandList::Commands for a list of supported commands.
+    def command_list(response_type=nil,&commands)
+      @mutex.synchronize do
+        begin
+          socket.puts "command_list_begin"
+          CommandList::Commands.new(self).instance_eval(&commands)
+          socket.puts "command_list_end"
+
+          # clear the response from the socket, even if we will not parse it
+          response = handle_server_response || ""
+
+          case response_type
+          when :values    then response.lines.map{ |line| parse_line(line).last }
+          when :hash      then build_hash(response)
+          when :hashes    then build_response(:commandlist,response,true)
+          when :songs     then build_songs_list parse_response(:listallinfo,response)
+          when :playlists then build_playlists  parse_response(:listplaylists,response)
+          end
+        rescue Errno::EPIPE
+          reconnect
+          retry
+        end
+      end
+    end
+  end
+
+  class CommandList::Commands
+    def initialize(mpd)
+      @mpd = mpd
+    end
+
+    include MPD::Plugins::Controls
+    include MPD::Plugins::PlaybackOptions
+    include MPD::Plugins::Queue
+    include MPD::Plugins::Stickers
+    include MPD::Plugins::Database
+
+    # List all of the playlists in the database
+    def playlists
+      send_command(:listplaylists)
```

Comment:
All the playlist methods currently use `@mpd.send_command`. `MPD#send_command` includes  `handle_server_response` and `parse_response` which we can't have called immediately with a command list. I can get around this for all the other MPD methods because they just call `send_command` (without an explicit receiver) and I provide my own `send_command` implementation for `CommandList::Commands`.  

The only way that I can think to re-use the playlist objects would be to use an instance variable set as a flag when the command list was started and mutate the way `send_command` works everywhere under such circumstances. (I don't think I need to worry about tracking this state per thread, since multiple threads trying to use the same MPD instance would cause problems with the socket communication.)

I'd also have to modify `MPD::Playlist#songs` to handle the case when no result is returned from `send_command`, but I've already done that in a few other places as part of this patch.

Would you prefer that approach? Having `MPD#send_command` modify its behavior based on an instance variable tracking whether we're in a command list or not?

Does this feel better to you?